packages/toolpad-app/src/toolpad/DomLoader.tsx

```diff
@@ -276,6 +343,9 @@ export interface DomLoader {
   saving: boolean;
   unsavedChanges: number;
   saveError: string | null;
+  pendingHistoryDom: appDom.AppDom;
+  undo: appDom.AppDom[];
```

Comment:
Some nits:

* Perhaps it makes sense to already anticipate that we will take selection in the undo stack (and maybe even other things) and foresee a type that can be extended with that?

    ```suggestion
    undo: { dom: appDom.AppDom }[];
    ```

* very much a nit, but in terms of naming `undo` and `redo` sound like methods, I'd consider naming these `undoStack` and `redoStack` to better convey what they are

Reply:
> Perhaps it makes sense to already anticipate that we will take selection in the undo stack (and maybe even other things) and foresee a type that can be extended with that?

Hmm, I'd then rather wait for an actual implementation. Initially I was thinking that we should do selection state part of the dom itself. With the approach you are proposing I assume selection and action would happen at the same time, where in other apps selection itself is considered a separate undoable action. If however selection was part of the dom, then we wouldn't need to do any extra handling for the undo/redo

> very much a nit, but in terms of naming undo and redo sound like methods, I'd consider naming these undoStack and redoStack to better convey what they are

Makes sense 👍 just renamed them

Reply:
>  Initially I was thinking that we should do selection state part of the dom itself. 

The dom is what we persist on disk and what we regard as the shareable artifact containing a toolpad application. The selection is local state, it shouldn't be part of that.

> With the approach you are proposing I assume selection and action would happen at the same time

Selection would add an entry in the undo stack with the same dom but a different slection.

Reply:
Fine, regardless I'd leave this out of scope for this PR then as it's already quite big